Current dev env gets stuck trying to index multiple big map contents with same key

The indexer is stuck on a block with conflicting primary key on the same bulk update call.

This is a log excerpt to pinpoint the issue

```log
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: java.sql.BatchUpdateException: Batch entry 0 insert into "tezos"."big_map_contents" ("big_map_id","key","key_hash","value") values ('-1','0x0000432c659871│
14012fa7c7c76463e93ed92c085e68','exprujywNVPjaEYkhTMsHtmsR7jeJwEw4RheauJC9tj65TR1PaMN7o','Pair {} 7850'),('-1','0x000086b7990605548cb13db091c7a68a46a7aef3d0a2','expruMJ3MpDTTKCd3jWWGN1ubrFT3y3qbZRQ8QyfAa1X2JWQPS6knk','Pair { Elt 0x014b71eab378ed953d0e590f7e4b7f0e48cb43859a00 0 ; Elt 0x01d06a71a8decd5bdee1024309│
540846429f58dd6300 50 } 8465'),('-1','0x000085ef0c18b31983603d978a152de4cd61803db881','exprtfKNhZ1G8vMscchFjt1G1qww2P93VTLHMuhyThVYygZLdnRev2','Pair {} 5985'),('-1','0x000086b7990605548cb13db091c7a68a46a7aef3d0a2','expruMJ3MpDTTKCd3jWWGN1ubrFT3y3qbZRQ8QyfAa1X2JWQPS6knk','Pair { Elt 0x01d06a71a8decd5bdee10243095│
40846429f58dd6300 50 } 8515') on conflict ("big_map_id", "key") do update set "key_hash"=EXCLUDED."key_hash","value"=EXCLUDED."value" was aborted: ERROR: ON CONFLICT DO UPDATE command cannot affect row a second time                                                                                                 │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]:   Hint: Ensure that no rows proposed for insertion within the same command have duplicate constrained values.  Call getNextException to see other errors i│
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at org.postgresql.jdbc.BatchResultHandler.handleError(BatchResultHandler.java:148)                                                                    │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:481)                                                                       │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at org.postgresql.jdbc.PgPreparedStatement.executeBatch(PgPreparedStatement.java:1538)                                                                │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at com.zaxxer.hikari.pool.HikariProxyPreparedStatement.executeBatch(HikariProxyPreparedStatement.java)                                                │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at slick.jdbc.JdbcBackend$BaseSession.withPreparedStatement(JdbcBackend.scala:489)                                                                    │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)                                                                    │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2440)                                                         │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at com.zaxxer.hikari.pool.ProxyStatement.executeBatch(ProxyStatement.java:128)                                                                        │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at slick.jdbc.JdbcBackend$SessionDef.withPreparedStatement(JdbcBackend.scala:425)                                                                     │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at com.github.tminglei.slickpg.ExPostgresProfile$InsertActionComposerImpl$MultiInsertOrUpdateAction.nativeUpsert(ExPostgresProfile.scala:157)         │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: 01:02:47.832 [main] INFO                                           com.zaxxer.hikari.HikariDataSource   - lorre.db - Shutdown completed.                  │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at akka.actor.dungeon.Children.attachChild$(Children.scala:48)                                                                                        │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at akka.actor.LightArrayRevolverScheduler$TaskHolder.run(LightArrayRevolverScheduler.scala:337)                                                       │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at scala.collection.IterableLike.foreach(IterableLike.scala:74)                                                                                       │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at scala.collection.AbstractIterable.foreach(Iterable.scala:56)                                                                                       │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at akka.actor.ActorSystemImpl.stopScheduler(ActorSystem.scala:937)                                                                                    │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at akka.dispatch.BatchingExecutor$BlockableBatch.run(BatchingExecutor.scala:91)                                                                       │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: Exception in thread "main" tech.cryptonomic.conseil.tezos.TezosErrors$BlocksProcessingFailed                                                              │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at scala.concurrent.impl.Promise.$anonfun$transform$1(Promise.scala:33)                                                                               │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:55)                                                               │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at akka.dispatch.BatchingExecutor$BlockableBatch.$anonfun$run$1(BatchingExecutor.scala:91)                                                            │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:44)                                            │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: 01:02:47.930 [shutdownHook1] INFO                                  tech.cryptonomic.conseil.Lorre$      - Doing clean-up                                  │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at com.github.tminglei.slickpg.ExPostgresProfile$InsertActionComposerImpl$MultiInsertOrUpdateAction.$anonfun$nativeUpsert$1(ExPostgresProfile.scala:16│
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at slick.jdbc.JdbcBackend$SessionDef.withPreparedStatement(JdbcBackend.scala:425)                                                                     │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at slick.jdbc.JdbcActionComponent$InsertActionComposerImpl.preparedInsert(JdbcActionComponent.scala:511)                                              │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: Caused by: org.postgresql.util.PSQLException: ERROR: ON CONFLICT DO UPDATE command cannot affect row a second time                                        │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2183)                                                               │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011... 20 common frames omitted                                                                                                                          │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]:   Hint: Ensure that no rows proposed for insertion within the same command have duplicate constrained values.  Call getNextException to see other errors i│
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at org.postgresql.jdbc.BatchResultHandler.handleError(BatchResultHandler.java:148)                                                                    │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at slick.jdbc.JdbcActionComponent$InsertActionComposerImpl.preparedInsert(JdbcActionComponent.scala:511)                                              │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at slick.jdbc.JdbcActionComponent$SimpleJdbcProfileAction.run(JdbcActionComponent.scala:28)                                                           │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: Caused by: org.postgresql.util.PSQLException: ERROR: ON CONFLICT DO UPDATE command cannot affect row a second time                                        │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at akka.stream.ActorMaterializer$.apply(ActorMaterializer.scala:65)                                                                                   │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at akka.actor.LightArrayRevolverScheduler.close(LightArrayRevolverScheduler.scala:140)                                                                │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at scala.concurrent.Future.$anonfun$andThen$1(Future.scala:536)                                                                                       │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at scala.concurrent.impl.Promise.liftedTree1$1(Promise.scala:33)                                                                                      │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at akka.dispatch.BatchingExecutor$AbstractBatch.processBatch(BatchingExecutor.scala:55)                                                               │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)                                                                                  │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)                                                                                  │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: 01:02:47.780 [lorre-system-akka.actor.default-dispatcher-3] ERROR  tech.cryptonomic.conseil.Lorre$      - Could not write blocks or accounts checkpoints t│
o the database.
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at org.postgresql.jdbc.PgStatement.executeBatch(PgStatement.java:840)                                                                                 │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at com.github.tminglei.slickpg.ExPostgresProfile$InsertActionComposerImpl$MultiInsertOrUpdateAction.nativeUpsert(ExPostgresProfile.scala:157)         │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at com.github.tminglei.slickpg.ExPostgresProfile$InsertActionComposerImpl$MultiInsertOrUpdateAction.run(ExPostgresProfile.scala:138)                  │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at slick.jdbc.JdbcActionComponent$SimpleJdbcProfileAction.run(JdbcActionComponent.scala:28)                                                           │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at slick.jdbc.JdbcActionComponent$SimpleJdbcProfileAction.run(JdbcActionComponent.scala:25)                                                           │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at java.lang.Thread.run(Thread.java:748)                                                                                                              │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2184)                                                               │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:481)                                                                       │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at org.postgresql.jdbc.PgPreparedStatement.executeBatch(PgPreparedStatement.java:1538)                                                                │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at slick.jdbc.JdbcBackend$SessionDef.withPreparedStatement$(JdbcBackend.scala:420)                                                                    │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at slick.jdbc.JdbcActionComponent$SimpleJdbcProfileAction.run(JdbcActionComponent.scala:25)                                                           │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)                                                                    │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]:   Hint: Ensure that no rows proposed for insertion within the same command have duplicate constrained values.                                             │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2440)                                                         │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: [ERROR] [05/16/2020 01:02:47.896] [lorre-system-akka.actor.default-dispatcher-3] [akka.actor.ActorSystemImpl(lorre-system)] exception while executing time│
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at akka.actor.ActorCell.attachChild(ActorCell.scala:429)                                                                                              │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at akka.stream.ActorMaterializer$.actorOfStreamSupervisor(ActorMaterializer.scala:73)                                                                 │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at akka.stream.ActorMaterializer$.apply(ActorMaterializer.scala:42)                                                                                   │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at tech.cryptonomic.conseil.Lorre$.$anonfun$new$3(Lorre.scala:112)                                                                                    │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at akka.actor.LightArrayRevolverScheduler$$anon$1$$anon$2.run(LightArrayRevolverScheduler.scala:101)                                                  │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at akka.actor.LightArrayRevolverScheduler.$anonfun$close$1$adapted(LightArrayRevolverScheduler.scala:140)                                             │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at akka.actor.ActorSystemImpl$$anon$3.run(ActorSystem.scala:892)                                                                                      │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at akka.actor.ActorSystemImpl$TerminationCallbacks$$anonfun$addRec$1$1.applyOrElse(ActorSystem.scala:1068)                                            │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:64)                                                                                       │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)                                                                             │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:85)                                                                             │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(ForkJoinExecutorConfigurator.scala:44)                                            │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)                                                                              │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: 01:02:47.917 [main] INFO                                           tech.cryptonomic.conseil.Lorre$      - All things closed                               │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at tech.cryptonomic.conseil.Lorre$.$anonfun$processTezosBlocks$42(Lorre.scala:502)                                                                    │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.java:23)                                                                             │
May 15th 2020, 21:02:47.000     spinner2        conseil-dev:2020-march-release-21-147   May 16 01:02:47 spinner2 conseil-dev:2020-march-release-21-147[6804]: #011at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:85)                                                                             │
```

--- a/conseil-api/src/test/resources/logback-test.xml
+++ b/conseil-api/src/test/resources/logback-test.xml
@@ -10,8 +10,6 @@
     </appender>
 
     <logger name="com.base22" level="TRACE"/>
-    <logger name="tech.cryptonomic.conseil.tezos.BigMapsOperations" level="DEBUG"/>
-    <logger name="tech.cryptonomic.conseil.tezos.michelson.contracts.TNSContracts" level="DEBUG"/>
 
     <root level="OFF">
         <appender-ref ref="STDOUT" />

--- a/conseil-common/src/main/scala/tech/cryptonomic/conseil/common/tezos/TezosOptics.scala
+++ b/conseil-common/src/main/scala/tech/cryptonomic/conseil/common/tezos/TezosOptics.scala
@@ -73,10 +73,10 @@ object TezosOptics {
     val setBalances: List[BalanceUpdate] => Block => Block = blockBalances set _
 
     /** functions to operate on all big maps copy diffs within a block */
-    val readBigMapDiffCopy = blockOperationsGroup composeTraversal groupOperations composeTraversal Operations.operationBigMapDiffCopy
+    val readBigMapDiffCopy = blockOperationsGroup composeTraversal groupOperations composeTraversal Operations.overOperationBigMapDiffCopy
 
     /** functions to operate on all big maps remove diffs within a block */
-    val readBigMapDiffRemove = blockOperationsGroup composeTraversal groupOperations composeTraversal Operations.operationBigMapDiffRemove
+    val readBigMapDiffRemove = blockOperationsGroup composeTraversal groupOperations composeTraversal Operations.overOperationBigMapDiffRemove
 
     /**  Utility extractor that collects, for a block, both operations and internal operations results, grouped
       * in a form more amenable to processing
@@ -98,6 +98,29 @@ object TezosOptics {
         }
         group -> (group.contents, internal)
       }.toMap
+
+    /** Extracts all operations, primary and internal, and pre-order traverse
+      * the two-level structure to try and preserve the original intended sequence,
+      * foregoing the grouping structure
+      */
+    def extractOperationsSequence(
+        block: Block
+    ): List[Either[Operation, InternalOperationResults.InternalOperationResult]] =
+      for {
+        group <- block.operationGroups
+        op <- group.contents
+        all <- op match {
+          case r: Reveal =>
+            Left(op) :: r.metadata.internal_operation_results.toList.flatten.map(Right(_))
+          case t: Transaction =>
+            Left(op) :: t.metadata.internal_operation_results.toList.flatten.map(Right(_))
+          case o: Origination =>
+            Left(op) :: o.metadata.internal_operation_results.toList.flatten.map(Right(_))
+          case d: Delegation =>
+            Left(op) :: d.metadata.internal_operation_results.toList.flatten.map(Right(_))
+          case _ => List(Left(op))
+        }
+      } yield all
 
     //Note, cycle 0 starts at the level 2 block
     def extractCycle(block: Block): Option[Int] =
@@ -122,100 +145,94 @@ object TezosOptics {
 
   object Operations {
     import tech.cryptonomic.conseil.common.tezos.TezosTypes.InternalOperationResults.{
-      Transaction => InternalTransaction
+      Transaction => InternalTransaction,
+      Origination => InternalOrigination
     }
     import tech.cryptonomic.conseil.common.tezos.TezosTypes.OperationResult.Status
 
-    val selectOrigination = GenPrism[Operation, Origination]
-    val originationResult = GenLens[Origination](_.metadata.operation_result)
-    val originationBigMapDiffs =
+    val whenOrigination = GenPrism[Operation, Origination]
+    val onOriginationResult = GenLens[Origination](_.metadata.operation_result)
+    val onOriginationBigMapDiffs =
       Optional[OperationResult.Origination, List[Contract.CompatBigMapDiff]](_.big_map_diff)(
         diffs => result => result.copy(big_map_diff = diffs.some)
       )
 
-    val selectTransaction = GenPrism[Operation, Transaction]
-    val transactionResult = GenLens[Transaction](_.metadata.operation_result)
-    val transactionBigMapDiffs =
+    val whenTransaction = GenPrism[Operation, Transaction]
+    val onTransactionResult = GenLens[Transaction](_.metadata.operation_result)
+    val onTransactionBigMapDiffs =
       Optional[OperationResult.Transaction, List[Contract.CompatBigMapDiff]](_.big_map_diff)(
         diffs => result => result.copy(big_map_diff = diffs.some)
       )
 
-    val selectBigMapAlloc = left[Contract.BigMapDiff, Contract.Protocol4BigMapDiff] composePrism GenPrism[
+    val whenBigMapAlloc = left[Contract.BigMapDiff, Contract.Protocol4BigMapDiff] composePrism GenPrism[
             Contract.BigMapDiff,
             Contract.BigMapAlloc
           ]
 
-    val selectBigMapUpdate = left[Contract.BigMapDiff, Contract.Protocol4BigMapDiff] composePrism GenPrism[
+    val whenBigMapUpdate = left[Contract.BigMapDiff, Contract.Protocol4BigMapDiff] composePrism GenPrism[
             Contract.BigMapDiff,
             Contract.BigMapUpdate
           ]
 
-    val selectBigMapCopy = left[Contract.BigMapDiff, Contract.Protocol4BigMapDiff] composePrism GenPrism[
+    val whenBigMapCopy = left[Contract.BigMapDiff, Contract.Protocol4BigMapDiff] composePrism GenPrism[
             Contract.BigMapDiff,
             Contract.BigMapCopy
           ]
 
-    val selectBigMapRemove = left[Contract.BigMapDiff, Contract.Protocol4BigMapDiff] composePrism GenPrism[
+    val whenBigMapRemove = left[Contract.BigMapDiff, Contract.Protocol4BigMapDiff] composePrism GenPrism[
             Contract.BigMapDiff,
             Contract.BigMapRemove
           ]
 
-    val operationBigMapDiffAlloc =
-      selectOrigination composeLens
-          originationResult composeOptional
-          originationBigMapDiffs composeTraversal
-          (Traversal.fromTraverse[List, Contract.CompatBigMapDiff] composeOptional selectBigMapAlloc)
+    val overOperationBigMapDiffAlloc =
+      whenOrigination composeLens
+          onOriginationResult composeOptional
+          onOriginationBigMapDiffs composeTraversal
+          (Traversal.fromTraverse[List, Contract.CompatBigMapDiff] composeOptional whenBigMapAlloc)
 
-    val operationBigMapDiffUpdate =
-      selectTransaction composeLens
-          transactionResult composeOptional
-          transactionBigMapDiffs composeTraversal
-          (Traversal.fromTraverse[List, Contract.CompatBigMapDiff] composeOptional selectBigMapUpdate)
+    val overOperationBigMapDiffUpdate =
+      whenTransaction composeLens
+          onTransactionResult composeOptional
+          onTransactionBigMapDiffs composeTraversal
+          (Traversal.fromTraverse[List, Contract.CompatBigMapDiff] composeOptional whenBigMapUpdate)
 
-    val operationBigMapDiffCopy =
-      selectTransaction composeLens
-          transactionResult composeOptional
-          transactionBigMapDiffs composeTraversal
-          (Traversal.fromTraverse[List, Contract.CompatBigMapDiff] composeOptional selectBigMapCopy)
+    val overOperationBigMapDiffCopy =
+      whenTransaction composeLens
+          onTransactionResult composeOptional
+          onTransactionBigMapDiffs composeTraversal
+          (Traversal.fromTraverse[List, Contract.CompatBigMapDiff] composeOptional whenBigMapCopy)
 
-    val operationBigMapDiffRemove =
-      selectTransaction composeLens
-          transactionResult composeOptional
-          transactionBigMapDiffs composeTraversal
-          (Traversal.fromTraverse[List, Contract.CompatBigMapDiff] composeOptional selectBigMapRemove)
+    val overOperationBigMapDiffRemove =
+      whenTransaction composeLens
+          onTransactionResult composeOptional
+          onTransactionBigMapDiffs composeTraversal
+          (Traversal.fromTraverse[List, Contract.CompatBigMapDiff] composeOptional whenBigMapRemove)
 
     private def isApplied(status: String) = Status.parse(status).contains(Status.applied)
 
     def extractAppliedOriginationsResults(block: Block) = {
-      val (ops, intOps) = TezosOptics.Blocks.extractOperationsAlongWithInternalResults(block).values.unzip
+      val operationSequence = TezosOptics.Blocks.extractOperationsSequence(block).collect {
+        case Left(op: Origination) => op.metadata.operation_result
+        case Right(intOp: InternalOrigination) => intOp.result
+      }
 
-      val results =
-        (ops.toList.flatten.collect { case op: Origination => op.metadata.operation_result }) ++
-            (intOps.toList.flatten.collect { case intOp: InternalOperationResults.Origination => intOp.result })
-
-      results.filter(result => isApplied(result.status))
+      operationSequence.filter(result => isApplied(result.status))
     }
 
     def extractAppliedTransactionsResults(block: Block) = {
-      val (ops, intOps) = TezosOptics.Blocks.extractOperationsAlongWithInternalResults(block).values.unzip
+      val operationSequence = TezosOptics.Blocks.extractOperationsSequence(block).collect {
+        case Left(op: Transaction) => op.metadata.operation_result
+        case Right(intOp: InternalTransaction) => intOp.result
+      }
 
-      val results =
-        (ops.toList.flatten.collect { case op: Transaction => op.metadata.operation_result }) ++
-            (intOps.toList.flatten.collect { case intOp: InternalTransaction => intOp.result })
-
-      results.filter(result => isApplied(result.status))
+      operationSequence.filter(result => isApplied(result.status))
     }
 
-    def extractAppliedTransactions(block: Block): List[Either[Transaction, InternalTransaction]] = {
-      val (ops, intOps) = TezosOptics.Blocks.extractOperationsAlongWithInternalResults(block).values.unzip
-
-      (ops.toList.flatten.collect {
-        case op: Transaction if isApplied(op.metadata.operation_result.status) => Left(op)
-      }) ++
-        (intOps.toList.flatten.collect {
-          case intOp: InternalTransaction if isApplied(intOp.result.status) => Right(intOp)
-        })
-    }
+    def extractAppliedTransactions(block: Block): List[Either[Transaction, InternalTransaction]] =
+      TezosOptics.Blocks.extractOperationsSequence(block).collect {
+        case Left(op: Transaction) if isApplied(op.metadata.operation_result.status) => Left(op)
+        case Right(intOp: InternalTransaction) if isApplied(intOp.result.status) => Right(intOp)
+      }
 
   }
 

--- a/conseil-lorre/src/main/resources/logback.xml
+++ b/conseil-lorre/src/main/resources/logback.xml
@@ -31,7 +31,7 @@
 
     <logger name="com.base22" level="TRACE"/>
     <logger name="slick" level="INFO"/>
-    <logger name="tech.cryptonomic.conseil.common.tezos.BigMapsOperations" level="${LORRE_BIG_MAPS_LOG_LEVEL:-OFF}"/>
+    <logger name="tech.cryptonomic.conseil.indexer.tezos.bigmaps.BigMapsOperations" level="${LORRE_BIG_MAPS_LOG_LEVEL:-OFF}"/>
     <logger name="tech.cryptonomic.conseil.indexer.tezos.michelson.contracts.TNSContracts" level="${LORRE_TNS_LOG_LEVEL:-INFO}"/>
 
     <root level="INFO">

--- a/conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/bigmaps/BigMapsOperations.scala
+++ b/conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/bigmaps/BigMapsOperations.scala
@@ -7,7 +7,7 @@ import scala.concurrent.ExecutionContext
 import scala.collection.immutable.TreeMap
 import cats.implicits._
 import tech.cryptonomic.conseil.common.util.Conversion.Syntax._
-import tech.cryptonomic.conseil.indexer.tezos.michelson.contracts.TokenContracts
+import tech.cryptonomic.conseil.indexer.tezos.michelson.contracts.{TNSContract, TokenContracts}
 import tech.cryptonomic.conseil.common.tezos.TezosTypes.{
   Block,
   Contract,
@@ -25,7 +25,6 @@ import TezosOptics.Operations.{
   extractAppliedTransactions,
   extractAppliedTransactionsResults
 }
-import tech.cryptonomic.conseil.common.tezos.michelson.contracts.TNSContract
 
 /** Defines big-map-diffs specific handling, from block data extraction to database storage
   *

--- a/conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/bigmaps/BigMapsOperations.scala
+++ b/conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/bigmaps/BigMapsOperations.scala
@@ -4,6 +4,7 @@ import com.typesafe.scalalogging.LazyLogging
 import com.github.tminglei.slickpg.ExPostgresProfile
 
 import scala.concurrent.ExecutionContext
+import scala.collection.immutable.TreeMap
 import cats.implicits._
 import tech.cryptonomic.conseil.common.util.Conversion.Syntax._
 import tech.cryptonomic.conseil.indexer.tezos.michelson.contracts.TokenContracts
@@ -19,7 +20,12 @@ import tech.cryptonomic.conseil.common.tezos.TezosTypes.Contract.BigMapUpdate
 import tech.cryptonomic.conseil.common.tezos.Tables
 import tech.cryptonomic.conseil.common.tezos.Tables.{BigMapContentsRow, BigMapsRow, OriginatedAccountMapsRow}
 import tech.cryptonomic.conseil.common.tezos.TezosOptics
-import tech.cryptonomic.conseil.indexer.tezos.michelson.contracts.TNSContract
+import TezosOptics.Operations.{
+  extractAppliedOriginationsResults,
+  extractAppliedTransactions,
+  extractAppliedTransactionsResults
+}
+import tech.cryptonomic.conseil.common.tezos.michelson.contracts.TNSContract
 
 /** Defines big-map-diffs specific handling, from block data extraction to database storage
   *
@@ -34,43 +40,64 @@ case class BigMapsOperations[Profile <: ExPostgresProfile](profile: Profile) ext
     * @param ec needed to compose db operations
     * @return the count of records added
     */
-  def copyContent(blocks: List[Block])(implicit ec: ExecutionContext): DBIO[Int] = {
-    val copyDiffs = if (logger.underlying.isDebugEnabled()) {
-      val diffsPerBlock = blocks.map(b => b.data.hash.value -> TezosOptics.Blocks.readBigMapDiffCopy.getAll(b))
+  def copyContent(blocks: List[Block])(implicit ec: ExecutionContext): DBIO[Option[Int]] = {
+    val diffsPerBlock = blocks.map(b => b.data -> TezosOptics.Blocks.readBigMapDiffCopy.getAll(b))
+    if (logger.underlying.isDebugEnabled()) {
       diffsPerBlock.foreach {
-        case (hash, diffs) if diffs.nonEmpty =>
+        case (blockData, diffs) if diffs.nonEmpty =>
           logger.debug(
             "For block hash {}, I'm about to copy the following big maps data: \n\t{}",
-            hash.value,
+            blockData.hash.value,
             diffs.mkString(", ")
           )
         case _ =>
       }
-      diffsPerBlock.map(_._2).flatten
-    } else blocks.flatMap(TezosOptics.Blocks.readBigMapDiffCopy.getAll)
-
-    //need to load the sources and copy them with a new destination id
-    val contentCopies = copyDiffs.collect {
-      case Contract.BigMapCopy(_, Decimal(sourceId), Decimal(destinationId)) =>
-        Tables.BigMapContents
-          .filter(_.bigMapId === sourceId)
-          .map(it => (destinationId, it.key, it.keyHash, it.operationGroupId, it.value))
-          .result
-          .map(rows => rows.map(BigMapContentsRow.tupled).toList)
     }
 
-    DBIO
-      .sequence(contentCopies)
-      .flatMap { updateRows =>
-        val copies = updateRows.flatten
-        logger.info(
-          "{} big maps will be copied in the db.",
-          if (copies.nonEmpty) s"A total of ${copies.size}"
-          else "No"
-        )
-        Tables.BigMapContents.insertOrUpdateAll(copies)
+    /* we load the sources and copy them with a new destination id
+     * collecting relevant diffs per block level
+     * What we get out is a sequence of queries whose results are the new rows
+     * to write back to db, sorted by growing level
+     */
+    val sortedQueries = {
+      val copyDataByLevel = diffsPerBlock.map {
+        case (blockData, diffs) =>
+          val queries = diffs.collect {
+            case Contract.BigMapCopy(_, Decimal(sourceId), Decimal(destinationId)) =>
+              Tables.BigMapContents
+                .filter(_.bigMapId === sourceId)
+                .map(it => (destinationId, it.key, it.keyHash, it.operationGroupId, it.value))
+                .result
+                .headOption
+          }
+          blockData.header.level -> DBIO.sequence(queries).map(_.flatten)
+
       }
-      .map(_.sum)
+      TreeMap(copyDataByLevel: _*)
+    }
+
+    def dedup(contents: List[BigMapContentsRow]) =
+      contents
+        .groupBy(row => (row.bigMapId, row.key))
+        .values
+        .flatMap(_.lastOption)
+
+    val writesByLevel = sortedQueries.values.map(
+      readAction =>
+        readAction.flatMap { updateData =>
+          val rowsToWrite = dedup(updateData.map(BigMapContentsRow.tupled))
+          Tables.BigMapContents.insertOrUpdateAll(rowsToWrite)
+        }
+    )
+
+    DBIO.sequence(writesByLevel).map { upserts =>
+      val all = upserts.fold(Some(0))(_ |+| _)
+      logger.info(
+        "{} big maps will be actually copied in the db.",
+        all.fold("An unspecified number of")(String.valueOf)
+      )
+      all
+    }
 
   }
 
@@ -118,7 +145,6 @@ case class BigMapsOperations[Profile <: ExPostgresProfile](profile: Profile) ext
     * @return the count of records added, if available from the underlying db-engine
     */
   def saveMaps(blocks: List[Block]): DBIO[Option[Int]] = {
-    import TezosOptics.Operations.extractAppliedOriginationsResults
 
     val diffsPerBlock = blocks.flatMap(
       b =>
@@ -161,7 +187,6 @@ case class BigMapsOperations[Profile <: ExPostgresProfile](profile: Profile) ext
     * @return the count of records added, if available from the underlying db-engine
     */
   def upsertContent(blocks: List[Block]): DBIO[Option[Int]] = {
-    import TezosOptics.Operations.extractAppliedTransactionsResults
 
     val diffsPerBlock = blocks.flatMap(
       b =>
@@ -198,8 +223,8 @@ case class BigMapsOperations[Profile <: ExPostgresProfile](profile: Profile) ext
       .foldLeft(Map.empty[(BigDecimal, String), BigMapContentsRow]) {
         case (collected, block) =>
           val seen = collected.keySet
-          val rows = blocks
-            .flatMap(b => rowsPerBlock.getOrElse(b.data.hash, List.empty))
+          val rows = rowsPerBlock
+            .getOrElse(block.data.hash, List.empty)
             .filterNot(row => seen((row.bigMapId, row.key)))
           collected ++ rows.map(row => (row.bigMapId, row.key) -> row)
       }
@@ -219,7 +244,6 @@ case class BigMapsOperations[Profile <: ExPostgresProfile](profile: Profile) ext
     * @return the count of records added, if available from the underlying db-engine
     */
   def saveContractOrigin(blocks: List[Block]): DBIO[List[OriginatedAccountMapsRow]] = {
-    import TezosOptics.Operations.extractAppliedOriginationsResults
 
     val diffsPerBlock = blocks.flatMap(
       b =>
@@ -310,7 +334,6 @@ case class BigMapsOperations[Profile <: ExPostgresProfile](profile: Profile) ext
       blocks: List[Block]
   )(implicit ec: ExecutionContext, tokenContracts: TokenContracts): DBIO[Option[Int]] = {
     import slickeffect.implicits._
-    import TezosOptics.Operations.extractAppliedTransactions
 
     val toSql = (zdt: java.time.ZonedDateTime) => java.sql.Timestamp.from(zdt.toInstant)
 

--- a/conseil-lorre/src/test/scala/tech/cryptonomic/conseil/indexer/tezos/TezosDatabaseOperationsTestFixtures.scala
+++ b/conseil-lorre/src/test/scala/tech/cryptonomic/conseil/indexer/tezos/TezosDatabaseOperationsTestFixtures.scala
@@ -683,14 +683,14 @@ trait TezosDatabaseOperationsTestFixtures extends RandomGenerationKit {
       * In this case the generation needs happen only for Originations
       */
     def updateOperationsWithBigMapAllocation(
-        diffGenerate: PartialFunction[Operation, Contract.BigMapAlloc]
+        diffGenerate: PartialFunction[Operation, List[Contract.BigMapAlloc]]
     ): List[Operation] => List[Operation] = {
       import tech.cryptonomic.conseil.common.tezos.TezosOptics.Operations._
 
       //applies to the alloc diffs nested within originations
       updateOperationsToBigMapDiff[Contract.BigMapAlloc](
         diffGenerate,
-        selectOrigination composeLens originationResult composeOptional originationBigMapDiffs
+        whenOrigination composeLens onOriginationResult composeOptional onOriginationBigMapDiffs
       )
     }
 
@@ -701,14 +701,14 @@ trait TezosDatabaseOperationsTestFixtures extends RandomGenerationKit {
       * In this case the generation needs happen only for Transacions
       */
     def updateOperationsWithBigMapUpdate(
-        diffGenerate: PartialFunction[Operation, Contract.BigMapUpdate]
+        diffGenerate: PartialFunction[Operation, List[Contract.BigMapUpdate]]
     ): List[Operation] => List[Operation] = {
       import tech.cryptonomic.conseil.common.tezos.TezosOptics.Operations._
 
       //applies to the update diffs nested within transactions
       updateOperationsToBigMapDiff[Contract.BigMapUpdate](
         diffGenerate,
-        selectTransaction composeLens transactionResult composeOptional transactionBigMapDiffs
+        whenTransaction composeLens onTransactionResult composeOptional onTransactionBigMapDiffs
       )
     }
 
@@ -719,14 +719,14 @@ trait TezosDatabaseOperationsTestFixtures extends RandomGenerationKit {
       * In this case the generation needs happen only for Transactions
       */
     def updateOperationsWithBigMapCopy(
-        diffGenerate: PartialFunction[Operation, Contract.BigMapCopy]
+        diffGenerate: PartialFunction[Operation, List[Contract.BigMapCopy]]
     ): List[Operation] => List[Operation] = {
       import tech.cryptonomic.conseil.common.tezos.TezosOptics.Operations._
 
       //applies to the update diffs nested within transactions
       updateOperationsToBigMapDiff[Contract.BigMapCopy](
         diffGenerate,
-        selectTransaction composeLens transactionResult composeOptional transactionBigMapDiffs
+        whenTransaction composeLens onTransactionResult composeOptional onTransactionBigMapDiffs
       )
     }
 
@@ -737,14 +737,14 @@ trait TezosDatabaseOperationsTestFixtures extends RandomGenerationKit {
       * In this case the generation needs happen only for Transacions
       */
     def updateOperationsWithBigMapRemove(
-        diffGenerate: PartialFunction[Operation, Contract.BigMapRemove]
+        diffGenerate: PartialFunction[Operation, List[Contract.BigMapRemove]]
     ): List[Operation] => List[Operation] = {
       import tech.cryptonomic.conseil.common.tezos.TezosOptics.Operations._
 
       //applies to the update diffs nested within transactions
       updateOperationsToBigMapDiff[Contract.BigMapRemove](
         diffGenerate,
-        selectTransaction composeLens transactionResult composeOptional transactionBigMapDiffs
+        whenTransaction composeLens onTransactionResult composeOptional onTransactionBigMapDiffs
       )
     }
 
@@ -757,17 +757,17 @@ trait TezosDatabaseOperationsTestFixtures extends RandomGenerationKit {
      * @return the updated operations
      */
     private def updateOperationsToBigMapDiff[Diff <: Contract.BigMapDiff](
-        diffGenerate: PartialFunction[Operation, Diff],
+        diffGenerate: PartialFunction[Operation, List[Diff]],
         diffSetter: Optional[Operation, List[Contract.CompatBigMapDiff]]
     )(operations: List[Operation]): List[Operation] =
       operations.map { op =>
         // applies the function to create a possible value to set on each individual operation
-        val maybeDiff = PartialFunction.condOpt(op)(diffGenerate)
+        val maybeDiffs = PartialFunction.condOpt(op)(diffGenerate)
 
         //sets the diff value list in the optional field of the operation if there's something, or returns the unchanged operation
-        maybeDiff.fold(op) { diff =>
-          val diffFieldValue = Left(diff) //look at how CompatBigMapDiff is defined
-          diffSetter.modify(diffs => diffFieldValue :: diffs)(op)
+        maybeDiffs.fold(op) { diffs =>
+          val diffFieldValues = diffs.map(Left(_)) //look at how CompatBigMapDiff is defined
+          diffSetter.modify(diffs => diffFieldValues ::: diffs)(op)
         }
       }
 

--- a/conseil-lorre/src/test/scala/tech/cryptonomic/conseil/indexer/tezos/bigmaps/BigMapsOperationsTest.scala
+++ b/conseil-lorre/src/test/scala/tech/cryptonomic/conseil/indexer/tezos/bigmaps/BigMapsOperationsTest.scala
@@ -8,7 +8,12 @@ import slick.jdbc.PostgresProfile.api._
 import tech.cryptonomic.conseil.common.sql.CustomProfileExtension
 import tech.cryptonomic.conseil.common.testkit.InMemoryDatabase
 import tech.cryptonomic.conseil.common.testkit.util.RandomSeed
-import tech.cryptonomic.conseil.common.tezos.Tables.{BigMapContentsRow, BigMapsRow, OriginatedAccountMapsRow, TokenBalancesRow}
+import tech.cryptonomic.conseil.common.tezos.Tables.{
+  BigMapContentsRow,
+  BigMapsRow,
+  OriginatedAccountMapsRow,
+  TokenBalancesRow
+}
 import tech.cryptonomic.conseil.common.tezos.{Tables, TezosTypes}
 import tech.cryptonomic.conseil.common.tezos.TezosTypes._
 import tech.cryptonomic.conseil.indexer.tezos.michelson.contracts.TokenContracts
@@ -44,7 +49,7 @@ class BigMapsOperationsTest
               big_map = Decimal(1),
               key_type = Micheline("""{"prim":"address"}"""),
               value_type = Micheline("""{"prim":"nat"}""")
-            )
+            ) :: Nil
         }
 
         val block = generateSingleBlock(1, testReferenceDateTime)
@@ -85,7 +90,7 @@ class BigMapsOperationsTest
               key = Micheline("""{"bytes":"0000b2e19a9e74440d86c59f13dab8a18ff873e889ea"}"""),
               key_hash = ScriptId("exprv6UsC1sN3Fk2XfgcJCL8NCerP5rCGy1PRESZAqr7L2JdzX55EN"),
               value = Some(Micheline("""{"prim":"Pair", "args": [{"int":"20"},[]]}"""))
-            )
+            ) :: Nil
         }
 
         //we need this to be referred as a FK from the content record
@@ -138,7 +143,7 @@ class BigMapsOperationsTest
               big_map = Decimal(1),
               key_type = Micheline("""{"prim":"address"}"""),
               value_type = Micheline("""{"prim":"nat"}""")
-            )
+            ) :: Nil
         }
 
         val block = generateSingleBlock(1, testReferenceDateTime)
@@ -190,7 +195,7 @@ class BigMapsOperationsTest
               key = Micheline("""{"bytes":"0000a8d45bdc966ddaaac83188a1e1c1fde2a3e05e5c"}"""),
               key_hash = ScriptId("exprvKTBQDAyXTMRc36TsLBsj9y5GXo1PD529MfF8zDV1pVzNNgehs"),
               value = Some(Micheline("""{"prim":"Pair", "args": [{"int":"50"},[]]}"""))
-            )
+            ) :: Nil
         }
 
         //we're gonna direct any sample transaction to the token contract, referring to a test account
@@ -282,7 +287,7 @@ class BigMapsOperationsTest
               key = Micheline("""{"bytes":"0000b2e19a9e74440d86c59f13dab8a18ff873e889ea"}"""),
               key_hash = ScriptId("exprv6UsC1sN3Fk2XfgcJCL8NCerP5rCGy1PRESZAqr7L2JdzX55EN"),
               value = Some(Micheline("""{"prim":"Pair", "args": [{"int":"50"},[]]}"""))
-            )
+            ) :: Nil
         }
 
         val block = generateSingleBlock(1, testReferenceDateTime)
@@ -309,6 +314,85 @@ class BigMapsOperationsTest
             value = Some("Pair 50 {}")
           )
         )
+      }
+
+      "update with only the latest big map content for diffs having the same target map and key: Issue#807" in {
+        //given
+        implicit val randomSeed = RandomSeed(testReferenceTimestamp.getTime)
+
+        val initialBigMap = BigMapsRow(
+          bigMapId = BigDecimal(1),
+          keyType = Some("address"),
+          valueType = Some("nat")
+        )
+        val initialBigMapContent = BigMapContentsRow(
+          bigMapId = BigDecimal(1),
+          key = "0x0000b2e19a9e74440d86c59f13dab8a18ff873e889ea",
+          keyHash = Some("exprv6UsC1sN3Fk2XfgcJCL8NCerP5rCGy1PRESZAqr7L2JdzX55EN"),
+          value = Some("Pair 20 {}")
+        )
+
+        val populate = for {
+          mapAdded <- Tables.BigMaps += initialBigMap
+          contentAdded <- Tables.BigMapContents += initialBigMapContent
+        } yield (mapAdded, contentAdded)
+
+        dbHandler.run(populate).futureValue shouldEqual ((1, 1))
+
+        //here we want to generate multiple updates in different blocks
+        def updatesMap(values: List[Int]): ListTransf[Operation] = Operations.updateOperationsWithBigMapUpdate {
+          case (_: Transaction) =>
+            values.map { value =>
+              Contract.BigMapUpdate(
+                action = "udpate",
+                big_map = Decimal(1),
+                key = Micheline("""{"bytes":"0000b2e19a9e74440d86c59f13dab8a18ff873e889ea"}"""),
+                key_hash = ScriptId("exprv6UsC1sN3Fk2XfgcJCL8NCerP5rCGy1PRESZAqr7L2JdzX55EN"),
+                value = Some(Micheline(s"""{"prim":"Pair", "args": [{"int":"$value"},[]]}"""))
+              )
+            }
+        }
+
+        /* we make two blocks and define which are the big map values to update
+         * therefore adding the diff to the corresponding operations
+         */
+        val blocks = generateBlocks(2, testReferenceDateTime).drop(1)
+
+        val operationsWithDiffs: List[List[OperationsGroup]] = blocks.map { block =>
+          val sampleOperations = generateOperationGroup(block, generateOperations = true)
+          val updateValue = block.data.header.level match {
+            case 1 => List(10) //block-lvl 1 will carry one update
+            case 2 => List(20, 50) //block-lvl 2 will carry both updates
+            case _ => List.empty
+          }
+          sampleOperations.copy(
+            contents = updatesMap(values = updateValue)(sampleOperations.contents)
+          ) :: Nil
+        }
+
+        val blocksToSave =
+          blocks.zip(operationsWithDiffs).map {
+            case (block, ops) => block.copy(operationGroups = ops)
+          }
+
+        //when
+        val writeAndGetRows = sut.upsertContent(blocksToSave) andThen Tables.BigMapContents.result
+
+        val contents = dbHandler.run(writeAndGetRows.transactionally).futureValue
+
+        //then
+        contents.size shouldBe 1
+
+        // we expect only the latest update to take effect
+        contents(0) should matchTo(
+          BigMapContentsRow(
+            bigMapId = BigDecimal(1),
+            key = "0x0000b2e19a9e74440d86c59f13dab8a18ff873e889ea",
+            keyHash = Some("exprv6UsC1sN3Fk2XfgcJCL8NCerP5rCGy1PRESZAqr7L2JdzX55EN"),
+            value = Some("Pair 50 {}")
+          )
+        )
+
       }
 
       "copy big map contents for diffs contained in a list of blocks" in {
@@ -350,7 +434,7 @@ class BigMapsOperationsTest
               action = "copy",
               source_big_map = Decimal(1),
               destination_big_map = Decimal(2)
-            )
+            ) :: Nil
         }
 
         val block = generateSingleBlock(1, testReferenceDateTime)
@@ -384,6 +468,105 @@ class BigMapsOperationsTest
           )
         )
 
+      }
+
+      "copy only the latest big map content for diffs copying to the same target map and key: Issue#807" in {
+        //given
+        implicit val randomSeed = RandomSeed(testReferenceTimestamp.getTime)
+
+        //we need 3 pre-existing big maps to transafer content between
+        val initialBigMaps =
+          1 :: 2 :: 3 :: Nil map (
+                  i =>
+                    BigMapsRow(
+                      bigMapId = BigDecimal(i),
+                      keyType = None,
+                      valueType = None
+                    )
+                )
+
+        //the content to copy
+        val initialBigMapsContent = List(
+          BigMapContentsRow(
+            bigMapId = BigDecimal(1),
+            key = "0x0000b2e19a9e74440d86c59f13dab8a18ff873e889ea",
+            keyHash = Some("exprv6UsC1sN3Fk2XfgcJCL8NCerP5rCGy1PRESZAqr7L2JdzX55EN"),
+            value = Some("Pair 10 {}")
+          ),
+          BigMapContentsRow(
+            bigMapId = BigDecimal(2),
+            key = "0x0000b2e19a9e74440d86c59f13dab8a18ff873e889ea",
+            keyHash = Some("exprv6UsC1sN3Fk2XfgcJCL8NCerP5rCGy1PRESZAqr7L2JdzX55EN"),
+            value = Some("Pair 20 {}")
+          )
+        )
+
+        //store the data
+        val populate = for {
+          maps <- Tables.BigMaps ++= initialBigMaps
+          contents <- Tables.BigMapContents ++= initialBigMapsContent
+        } yield (maps, contents)
+
+        dbHandler.run(populate.transactionally).futureValue shouldEqual ((Some(3), Some(2)))
+
+        //we want to copy content between the first two maps and the third map, on the same key
+        def copyMap(from: List[Int], to: Int): ListTransf[Operation] = Operations.updateOperationsWithBigMapCopy {
+          case (_: Transaction) =>
+            from.map { sourceId =>
+              Contract.BigMapCopy(
+                action = "copy",
+                source_big_map = Decimal(sourceId),
+                destination_big_map = Decimal(to)
+              )
+
+            }
+        }
+
+        /* we make two blocks and define which are the big map ids to copy
+         * therefore adding the copy diff to the corresponding operations
+         */
+        val blocks = generateBlocks(2, testReferenceDateTime).drop(1)
+
+        val operationsWithDiffs: List[List[OperationsGroup]] = blocks.map { block =>
+          val sampleOperations = generateOperationGroup(block, generateOperations = true)
+          val sourceIdsToCopy = block.data.header.level match {
+            case 1 => List(1) //block-lvl 1 will copy contents from map 1
+            case 2 => List(1, 2) //block-lvl 2 will copy contents from map 1 and 2 both
+            case _ => List.empty
+          }
+          sampleOperations.copy(
+            contents = copyMap(from = sourceIdsToCopy, to = 3)(sampleOperations.contents)
+          ) :: Nil
+        }
+
+        val blocksToSave =
+          blocks.zip(operationsWithDiffs).map {
+            case (block, ops) => block.copy(operationGroups = ops)
+          }
+
+        //when
+        val writeAndGetRows = for {
+          _ <- sut.copyContent(blocksToSave)
+          maps <- Tables.BigMaps.result
+          contents <- Tables.BigMapContents.result
+        } yield (maps, contents)
+
+        val (maps, contents) = dbHandler.run(writeAndGetRows).futureValue
+
+        //then
+        maps.size shouldBe initialBigMaps.size
+
+        contents.size shouldBe 3
+
+        // we expect the destination map to only consider the very last copy applied
+        contents(2) should matchTo(
+          BigMapContentsRow(
+            bigMapId = BigDecimal(3),
+            key = "0x0000b2e19a9e74440d86c59f13dab8a18ff873e889ea",
+            keyHash = Some("exprv6UsC1sN3Fk2XfgcJCL8NCerP5rCGy1PRESZAqr7L2JdzX55EN"),
+            value = Some("Pair 20 {}")
+          )
+        )
       }
 
       "delete all data in selected big maps for diffs contained in a list of blocks" in {
@@ -437,7 +620,7 @@ class BigMapsOperationsTest
             Contract.BigMapRemove(
               action = "remove",
               big_map = Decimal(1)
-            )
+            ) :: Nil
         }
 
         val block = generateSingleBlock(1, testReferenceDateTime)

--- a/conseil-lorre/src/test/scala/tech/cryptonomic/conseil/indexer/tezos/bigmaps/BigMapsOperationsTest.scala
+++ b/conseil-lorre/src/test/scala/tech/cryptonomic/conseil/indexer/tezos/bigmaps/BigMapsOperationsTest.scala
@@ -375,8 +375,11 @@ class BigMapsOperationsTest
             case (block, ops) => block.copy(operationGroups = ops)
           }
 
+        //we change the order of how blocks come in
+        val reverted = blocksToSave.reverse
+
         //when
-        val writeAndGetRows = sut.upsertContent(blocksToSave) andThen Tables.BigMapContents.result
+        val writeAndGetRows = sut.upsertContent(reverted) andThen Tables.BigMapContents.result
 
         val contents = dbHandler.run(writeAndGetRows.transactionally).futureValue
 
@@ -544,9 +547,12 @@ class BigMapsOperationsTest
             case (block, ops) => block.copy(operationGroups = ops)
           }
 
+        //we change the order of how blocks come in
+        val reverted = blocksToSave.reverse
+
         //when
         val writeAndGetRows = for {
-          _ <- sut.copyContent(blocksToSave)
+          _ <- sut.copyContent(reverted)
           maps <- Tables.BigMaps.result
           contents <- Tables.BigMapContents.result
         } yield (maps, contents)

